### PR TITLE
Update gossipsub v1.1 spec with recommendations for network bootstrapping

### PR DESCRIPTION
**Pull request by vyzo**
The basic premise is bootstrapper PX instead of peer discovery services.

### CODE REVIEW

File: pubsub/gossipsub/gossipsub-v1.1.md [hunk 1]
```diff
@@ -626,3 +627,28 @@ In order counter spam that elicits responses and consumes resources, some measur
   penalized by the score function. A peer transmitting lots of spam will quickly get graylisted,
   reducing the surface of spam-induced computation (eg validation). The application can take
   further steps and blacklist the peer if the spam persists after the negative score decays.
+
+### Recommendations for Network Operators
+
+An important issue to consider when deploying gossipsub is the peer discovery mechanism,
+which must provide a secure way of discovering new peers.
+Prior to gossipsub v1.1, operators were required to utilize an external peer discovery
+mechanism; with gossipsub v1.1 this is now entirely optional and the network can bootstrap
```
Comment (raulk) [inline pubsub/gossipsub/gossipsub-v1.1.md:636]:
```suggestion
mechanism to locate peers meshing in particular topics; with gossipsub v1.1 this is now entirely optional and the network can bootstrap
```

File: pubsub/gossipsub/gossipsub-v1.1.md [hunk 2]
```diff
@@ -626,3 +627,28 @@ In order counter spam that elicits responses and consumes resources, some measur
...
+purely through a small set of network entry points (bootstrappers) by utilizing Peer Exchange.
```
Comment (raulk) [inline pubsub/gossipsub/gossipsub-v1.1.md:637]:
```suggestion
purely through a small set of network entry points (bootstrappers) by utilizing Peer Exchange. In other words, gossipsub 1.1 is now self-sufficient in this regard, as long as the node manages to find at least one peer meshing in the topic of interest.
```

File: pubsub/gossipsub/gossipsub-v1.1.md [hunk 3]
```diff
@@ -626,3 +627,28 @@ In order counter spam that elicits responses and consumes resources, some measur
...
+
+In order to successfully bootstrap the network without a discovery service, network operators
+should
+- Create and operate a set of bootstrapper nodes, that are known within client software.
+- The bootstrappers should be configured without a mesh (ie set `D=Dlo=Dhi=Dout=0`)
+  and with Peer Exchange enabled, utilizing Signed Peer Records.
```
Comment (raulk) [inline pubsub/gossipsub/gossipsub-v1.1.md:643]: We really need a proper spec for _signed peer records_ :-/

File: pubsub/gossipsub/gossipsub-v1.1.md (hunk 2 above)
Comment (raulk) [reply pubsub/gossipsub/gossipsub-v1.1.md:637]: ^^ is this correct, @vyzo?

File: pubsub/gossipsub/gossipsub-v1.1.md [hunk 4]
```diff
@@ -626,3 +627,28 @@ In order counter spam that elicits responses and consumes resources, some measur
...
+
+An important issue to consider when deploying gossipsub is the peer discovery mechanism,
+which must provide a secure way of discovering new peers.
+Prior to gossipsub v1.1, operators were required to utilize an external peer discovery
+mechanism; with gossipsub v1.1 this is now entirely optional and the network can bootstrap
+purely through a small set of network entry points (bootstrappers) by utilizing Peer Exchange.
+
+In order to successfully bootstrap the network without a discovery service, network operators
+should
+- Create and operate a set of bootstrapper nodes, that are known within client software.
```
Comment (raulk) [inline pubsub/gossipsub/gossipsub-v1.1.md:641]:
```suggestion
- Create and operate a set of stable bootstrapper nodes, whose addresses are known ahead of time by the application.
```

File: pubsub/gossipsub/gossipsub-v1.1.md [hunk 5]
```diff
@@ -626,3 +627,28 @@ In order counter spam that elicits responses and consumes resources, some measur
...
+- The bootstrappers should be configured without a mesh (ie set `D=Dlo=Dhi=Dout=0`)
+  and with Peer Exchange enabled, utilizing Signed Peer Records.
+- Client software should assign a high application specific score to the bootstrappers and
```
Comment (raulk) [inline pubsub/gossipsub/gossipsub-v1.1.md:644]:
```suggestion
- The application should assign a high application-specific score to the bootstrappers and
```

File: pubsub/gossipsub/gossipsub-v1.1.md [hunk 6]
```diff
@@ -626,3 +627,28 @@ In order counter spam that elicits responses and consumes resources, some measur
...
+  set the `AcceptPXThreshold` to a high enough value attainable only by the bootstrappers.
```
Comment (raulk) [inline pubsub/gossipsub/gossipsub-v1.1.md:645]:
```suggestion
  set `AcceptPXThreshold` to a high enough value attainable only by the bootstrappers.
```

File: pubsub/gossipsub/gossipsub-v1.1.md [hunk 7]
```diff
@@ -626,3 +627,28 @@ In order counter spam that elicits responses and consumes resources, some measur
...
+An important issue to consider when deploying gossipsub is the peer discovery mechanism,
+which must provide a secure way of discovering new peers.
+Prior to gossipsub v1.1, operators were required to utilize an external peer discovery
+mechanism; with gossipsub v1.1 this is now entirely optional and the network can bootstrap
+purely through a small set of network entry points (bootstrappers) by utilizing Peer Exchange.
+
+In order to successfully bootstrap the network without a discovery service, network operators
+should
+- Create and operate a set of bootstrapper nodes, that are known within client software.
+- The bootstrappers should be configured without a mesh (ie set `D=Dlo=Dhi=Dout=0`)
```
Comment (raulk) [inline pubsub/gossipsub/gossipsub-v1.1.md:642]: Can we add an option to the pubsub constructor in the go reference implementation that does this, e.g. `Bootstrapper()`.

File: pubsub/gossipsub/gossipsub-v1.1.md [hunk 8]
```diff
@@ -626,3 +627,28 @@ In order counter spam that elicits responses and consumes resources, some measur
...
+  and with Peer Exchange enabled, utilizing Signed Peer Records.
+- Client software should assign a high application specific score to the bootstrappers and
+  set the `AcceptPXThreshold` to a high enough value attainable only by the bootstrappers.
+
+In this manner, the bootstrappers act purely as gossip and peer exchange nodes that facilitate
+the formation and maintenance of the network.
+Note that the score function is still present in the bootstrappers, which ensures that invalid
+messages, colocation, and behavioural penalties apply to misbehaving nodes such that they do
+not receive PX or are advertised to the rest of the network.
+In addition, network operators may configure the application specific scoring function such
```
Comment (raulk) [inline pubsub/gossipsub/gossipsub-v1.1.md:652]:
```suggestion
In addition, network operators may configure the application-specific scoring function such
```

File: pubsub/gossipsub/gossipsub-v1.1.md (hunk 7 above)
Comment (vyzo) [reply pubsub/gossipsub/gossipsub-v1.1.md:642]: yes, we can do that.

File: pubsub/gossipsub/gossipsub-v1.1.md (hunk 2 above)
Comment (vyzo) [reply pubsub/gossipsub/gossipsub-v1.1.md:637]: yeah, it's fine. Perhaps not using the word "meshing" though, as the bootstrappers are recommended to use `D=0` so they have no mesh.

File: pubsub/gossipsub/gossipsub-v1.1.md (hunk 1 above)
Comment (vyzo) [reply pubsub/gossipsub/gossipsub-v1.1.md:636]: the word "meshing" is probably not entirely correct, i think we need "participating".

Comment (vyzo) [reply pubsub/gossipsub/gossipsub-v1.1.md:636]: I'll commit and fix the word by hand.

File: pubsub/gossipsub/gossipsub-v1.1.md (hunk 3 above)
Comment (yusefnapora) [reply pubsub/gossipsub/gossipsub-v1.1.md:643]: Very true. I can get that ready next week, I think. The implementation ended up tracking the RFC pretty closely, so it shouldn't take too much editing to make it into a spec.

**Pull request merged**